### ISSUE TITLE
[parley] Limit input text size to 2^30 bytes

### ISSUE THREAD

**Issue by dfrg** [linebender/parley#107]
1 GB of text per layout is more than generous. We're likely to see severe performance degradation long before reaching this limit anyway suggesting that applications wanting to support excessively large buffers will need to do some segmentation and caching.

Rationale: this lets us use `u32` instead of `usize` for internal indices which can significantly reduce memory requirements for internal data structures. Using 30 bits instead of 32 (or 31) allows us to occasionally steal a bit for other info (often incredibly useful) while still preserving the use of `MAX_SIZE + 1` as the exclusive end boundary.

**Comment by DJMcNab** [comment]
To be clear, are you wanting `MAX_SIZE + 1` to be $$2^{30}$$, or $$2^{30}-1$$? That is, do we have bits 30 and 31 available as "scratch", or just bit 31?

**Comment by dfrg** [comment]
Yes :) That’s my off by one error. I think we might as well reserve the 2 bits because 1 GB per layout seems to be plenty.

**Comment by xorgy** [comment]
I didn't find this issue and posted #408. Here's what I came up with (almost exactly the same conclusions as Chad 😂):

Laying out 2³² bytes of simple ASCII text with reasonable line lengths in Parley takes hundreds of gigabytes of memory†; and likely over an hour of time to complete; and the resulting Layout is almost unusable on account of its size.

There is almost no practical application for Parley attempting to lay out 4 GiB (or even 1 GiB) of text in a single layout; realistically nobody is going to wait half an hour to lay out a gigabyte of text (one time) just for the convenience of it being in a single layout, and be willing to spend something approaching a terabyte of memory on actually using the result of that layout.

Given that indexes, both relative and absolute, are littered throughout Parley; and given that layouts larger than a few megabytes even are not practical, let alone layouts exceeding 2³¹ bytes, I propose we look at setting a limit per layout of something like 2³⁰ bytes.
I suggest one less than 31 bits because it would allow us to use the most or least significant bit for other information; a simple (but maybe not that important) example being: cursor position being i32 with the least significant bit giving affinity, such that zero is the cursor at the start of the layout, `(len << 1) + 1` is the cursor at the end of the layout, and `cursor >> 1` is inherently a valid byte index in the layout (either the end or the start of a cluster).

Reducing the size of the index type would also improve the memory footprint of Parley quite a bit if done correctly, and maybe by reducing the theoretical maximum length of the text, we would add 20-30% to the _de facto_ maximum length of a layout in a given application (constrained by either time or memory).

And since every other length or offset in Parley pertains to one or more bytes in the input text, this means that indexes for lines, clusters, runs, etc. could also be shrunk to i32.

† The most I can do on my 128 GiB machines is 2²⁹ bytes, and then I run out of memory if I try to iterate on lines, so it's useless anyway.
